### PR TITLE
Add reports to rake for nil ethnicity data fix

### PR DESCRIPTION
**Pull request by mdkaraman**
ME-180666850

### CODE REVIEW

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 1]
```diff
@@ -1,39 +1,50 @@
...
-    # - check and update person ethnicity array
-    # - check and update applicant ethnicity array
+    # - check and update person ethnicity arrays
+    # - check and update applicant ethnicity arrays
 
     def find_person(person_hbx_id)
       Person.find_by(hbx_id: person_hbx_id)
     rescue StandardError => e
-      puts "Error finding person document - hbx_id: #{person_hbx_id}"
+      ["Error finding person document - hbx_id: #{person_hbx_id}"]
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:15]: this is not needed. you can revert this change. This method will return nil when there's no person with hbx id.

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 2]
```diff
@@ -1,39 +1,50 @@
...
+        person.ethnicity = ethnicities.compact
+        updated = person.save(validate: false)
+        Person.set_callback(:update, :after, :person_create_or_update_handler)
+        result << person.ethnicity.to_s
+        raise StandardError, "Failed to save person (hbx_id): #{person.hbx_id}" unless updated
+      end
+      result
     rescue StandardError => e
-      puts "Error updating peson ethnicity array - #{e}"
+      ["Error updating person ethnicity array - #{e}"]
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:31]: why this change?

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 3]
```diff
@@ -1,39 +1,50 @@
...
+      result = [application.hbx_id, primary, person.hbx_id, person.ethnicity.to_s]
+      if ENV['report_only'].nil?
+        Person.skip_callback(:update, :after, :person_create_or_update_handler)
+        person.ethnicity = ethnicities.compact
+        updated = person.save(validate: false)
+        Person.set_callback(:update, :after, :person_create_or_update_handler)
+        result << person.ethnicity.to_s
+        raise StandardError, "Failed to save person (hbx_id): #{person.hbx_id}" unless updated
+      end
+      result
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:29]: you don't have to return this `result`. it's not being used. 

And, if the only reason to have this is to get report only, can we add this as a different rake? I mean copy & paste it to a new rake it would be less confusing.

Also. the ideal way is - if the update failed you can simply print statement with why it' failed. We don't really need to raise a standard error. but it's fine as it's a report.

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 4]
```diff
@@ -1,39 +1,50 @@
...
+        applicant.ethnicity = ethnicities.compact
+        updated = applicant.save(validate: false)
+        ::FinancialAssistance::Applicant.set_callback(:update, :after, :propagate_applicant)
+        raise StandardError, "Failed to save applicant (person_hbx_id): #{applicant.person_hbx_id}" unless updated
+        result << applicant.ethnicity.to_s
+      end
+      result
     rescue StandardError => e
-      puts "Error updating applicant ethnicity array - #{e}"
+      ["Error updating applicant ethnicity array - #{e}"]
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:47]: same comment as above.

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 5]
```diff
@@ -1,39 +1,50 @@
...
-      @primary_hbx_ids << result
+      primary = application.primary_applicant ? application.primary_applicant.person_hbx_id : "no_primary_found"
+      result = [application.hbx_id, primary, applicant.person_hbx_id, applicant.ethnicity.to_s]
+      if ENV['report_only'].nil?
+        ::FinancialAssistance::Applicant.skip_callback(:update, :after, :propagate_applicant)
+        applicant.ethnicity = ethnicities.compact
+        updated = applicant.save(validate: false)
+        ::FinancialAssistance::Applicant.set_callback(:update, :after, :propagate_applicant)
+        raise StandardError, "Failed to save applicant (person_hbx_id): #{applicant.person_hbx_id}" unless updated
+        result << applicant.ethnicity.to_s
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:43]: same comment as above.

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 6]
```diff
@@ -43,32 +54,55 @@ namespace :migrations do
                        FinancialAssistance::Application.draft.where(hbx_id: application_hbx_id)
                      else
                        # Run on all ATP-ingested draft applications
-                       FinancialAssistance::Application.draft.not.where(transfer_id: nil)
+                       FinancialAssistance::Application.draft.where(:transfer_id.nin => [nil, ''])
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:57]: 👍

File: lib/tasks/fix_nil_ethnicity_arrays.rake [hunk 7]
```diff
@@ -43,32 +54,55 @@ namespace :migrations do
...
+      file_name = "#{Rails.root}/person_nil_ethnicity_#{timestamp}.csv"
+      FileUtils.touch(file_name)
+      CSV.open(file_name, 'w+', headers: true) do |csv|
+        headers = %w[APPLICATION_HBX_ID PRIMARY_HBX_ID PERSON_HBX_ID ETHNICITY]
+        headers << "UPDATED_ETHNICIY" unless ENV['report_only'].present?
+        csv << headers
+        applications.no_timeout.each do |application|
+          application.applicants.no_timeout.each do |applicant|
+            person = find_person(applicant.person_hbx_id)
+            if person.is_a? Array
```
Comment (nisanthyaganti9) [inline lib/tasks/fix_nil_ethnicity_arrays.rake:72]: you can simply do person.nil? / person.blank? and continue to next.

**Review by nisanthyaganti9 — CHANGES_REQUESTED** [review]

File: lib/tasks/fix_nil_ethnicity_arrays.rake (hunk 1 above)
Comment (mdkaraman) [reply lib/tasks/fix_nil_ethnicity_arrays.rake]: Person.find_by will raise a Mongoid::Errors::DocumentNotFound if it can't find a matching document.  I changed method to use Person.where, which will return nil if no match is found.

File: lib/tasks/fix_nil_ethnicity_arrays.rake (hunk 2 above)
Comment (mdkaraman) [reply lib/tasks/fix_nil_ethnicity_arrays.rake]: This was meant to log any encountered errors in the output .csv file.  I modified this so that errors are printed to the console and only successful updates are logged in the .csv

**Review by nisanthyaganti9 — APPROVED** [review]
